hadoop-ozone/s3gateway/src/test/java/org/apache/hadoop/ozone/s3/endpoint/TestObjectTaggingPut.java

```diff
@@ -0,0 +1,223 @@
+    ByteArrayInputStream body =
+        new ByteArrayInputStream("".getBytes(UTF_8));
+    objectEndpoint.setHeaders(headers);
+
+    objectEndpoint.put(BUCKET_NAME, KEY_NAME, 0, 1, null, null, body);
+  }
+
+  @Test
+  public void testPutObjectTaggingWithEmptyBody() throws Exception {
+    try {
+      objectEndpoint.put(BUCKET_NAME, KEY_NAME, 0, 1, null, "",
+          null);
+      fail();
+    } catch (OS3Exception ex) {
+      assertEquals(HTTP_BAD_REQUEST, ex.getHttpCode());
+      assertEquals(MALFORMED_XML.getCode(), ex.getCode());
+    }
+  }
+
+  @Test
+  public void testPutValidObjectTagging() throws Exception {
+    assertEquals(HTTP_OK, objectEndpoint.put(BUCKET_NAME, KEY_NAME, 0, 1,
+        null, "", twoTags()).getStatus());
+    OzoneKeyDetails keyDetails =
+        clientStub.getObjectStore().getS3Bucket(BUCKET_NAME).getKey(KEY_NAME);
+    assertEquals(2, keyDetails.getTags().size());
+    assertEquals("val1", keyDetails.getTags().get("tag1"));
+    assertEquals("val2", keyDetails.getTags().get("tag2"));
+  }
+
+  @Test
+  public void testPutInvalidObjectTagging() throws Exception {
+    testInvalidObjectTagging(this::emptyBody, HTTP_BAD_REQUEST, MALFORMED_XML.getCode());
+    testInvalidObjectTagging(this::invalidXmlStructure, HTTP_BAD_REQUEST, MALFORMED_XML.getCode());
+    testInvalidObjectTagging(this::noTagSet, HTTP_BAD_REQUEST, MALFORMED_XML.getCode());
+    testInvalidObjectTagging(this::emptyTags, HTTP_BAD_REQUEST, MALFORMED_XML.getCode());
+    testInvalidObjectTagging(this::tagKeyNotSpecified, HTTP_BAD_REQUEST, MALFORMED_XML.getCode());
+    testInvalidObjectTagging(this::tagValueNotSpecified, HTTP_BAD_REQUEST, MALFORMED_XML.getCode());
+  }
+
+  private void testInvalidObjectTagging(Supplier<InputStream> inputStream,
+                                        int expectedHttpCode, String expectedErrorCode) throws Exception {
+    try {
+      objectEndpoint.put(BUCKET_NAME, KEY_NAME, 0, 1, null, "",
+          inputStream.get());
+      fail("Expected an OS3Exception to be thrown");
+    } catch (OS3Exception ex) {
+      assertEquals(expectedHttpCode, ex.getHttpCode());
+      assertEquals(expectedErrorCode, ex.getCode());
+    }
+  }
+
+  @Test
+  public void testPutObjectTaggingNoKeyFound() throws Exception {
+    try {
+      objectEndpoint.put(BUCKET_NAME, "nonexistent", 0, 1,
+          null, "", twoTags());
+      fail("Expected an OS3Exception to be thrown");
+    } catch (OS3Exception ex) {
+      assertEquals(HTTP_NOT_FOUND, ex.getHttpCode());
+      assertEquals(NO_SUCH_KEY.getCode(), ex.getCode());
+    }
+  }
+
+  @Test
+  public void testPutObjectTaggingNoBucketFound() throws Exception {
+    try {
+      objectEndpoint.put("nonexistent", "nonexistent", 0, 1,
+          null, "", twoTags());
+      fail("Expected an OS3Exception to be thrown");
+    } catch (OS3Exception ex) {
+      assertEquals(HTTP_NOT_FOUND, ex.getHttpCode());
+      assertEquals(NO_SUCH_BUCKET.getCode(), ex.getCode());
+    }
+  }
```

Comment:
Could you add another test for NOT_IMPLEMENTED error code here and in TestObjectDeleteTagging.java

Reply:
Thanks for the review @SaketaChalamchala, updated PTAL.